src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+                TODO: Consider using a custom type for the handler to represent
+                      or document the interface to that callable.
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> DalMethodRequester:
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+
+class MethodMiddleware:
+    """
+    Base class used primarily for type checking and to provide
+    an interface for a method middleware.
+    """
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        pass
+
+
+def dal_resolver(ctx, dmr):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    return resolve(DalMethodRequest(ctx=ctx, path=dmr.path, args=None, kwargs=None))
+
+
+def resolve(request: DalMethodRequest):
+    """
+    This function resolves a DalMethodRequest to a service method
+    """
+    service_or_method = request.ctx.dal._services  # noqa
+    paths = list(request.path[:])
+    location = []
+    while paths:
+        path_segment = paths.pop(0)
+        location.append(path_segment)
+        # Third time through, service_or_method might be `None`, but we want
+        # to continue walking the path. Everything after the first missing
+        # service/method will be `(location, None)`.
+        if service_or_method:
+
+            # This if condition is handling the case of the first loop here.
+            # we cannot use attribute access on the dal directly because of how
+            # attribute access is deferred with DalMethodRequest objects.
+            if isinstance(service_or_method, dict):
+
+                # If the code being resolved has typo'd a service name, this
+                # could be returning something that is not a service.
+                service_or_method = service_or_method.get(path_segment.name)
+            else:
+                # second time through (and beyond), service or method is a real
+                # service or method, and we do not need to use a special case for
+                # finding the first service.
+                try:
+                    service_or_method = getattr(service_or_method, path_segment.name)
+                    if not service_or_method and callable(service_or_method):
+                        raise DalMethodError(request, path=tuple(location))
+
+                except KeyError:
+                    raise DalMethodError(request, path=tuple(location))
+        else:
+            raise DalMethodError(request, path=tuple(location))
```

Comment:
I had to read this a few times, but I think I understand what you're saying now. The gist of it is:

- polydatum should expose (somehow) a way to resolve these PathSegment chains.
- If the way polydatum does that is wrapped up entirely in a middleware, subclassing method middlewares wont' have a chance to do whatever they want with the PathSegment chains.  The nominal use case is erroring, but it may not always be what is wanted.
- impl could be in a function or a method on the default middleware, either way, it needs to be non-erroring so it can be used and the results of the resolve are left up to the brains of the middlewares.

Reply:
Yes. Other middleware should be able to walk the path, even if the path finds nothing at that segment. The middleware gets to decide what to do if there is nothing at that segment.